Xcode 16 SwiftUI.swiftinterface fails to compile with open source swiftc

### Description

The open source `swiftc` from a recent `main` fails to compile some of the tests of the test suite that use `SwiftUI` when the selected Xcode is one of the recent Xcode 16 betas.

Tests that fail:
- `Sema/SwiftUI/rdar104687668.swift`
- `Sema/SwiftUI/mixed_swiftui_and_multistatement_closures.swift`
- `Sema/SwiftUI/issue-56591.swift`
- `Sema/SwiftUI/issue-56479.swift`
- `Sema/SwiftUI/radar88971160.swift`
- `Sema/SwiftUI/builder_requirement_errors_should_not_shadow_errors_in_code.swift`
- `Sema/SwiftUI/issue-59716.swift`

The test report errors like the following:

```
<unknown>:0: error: fatal error encountered while in -verify mode
/Users/danielrodriguez/code/swift-source/swift/validation-test/Sema/SwiftUI/issue-59716.swift:4:8: error: unexpected error produced: failed to build module 'SwiftUI'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 6.0 effective-5.10 (swiftlang-6.0.0.5.20 clang-1600.0.22.6)', while this compiler is 'Apple Swift version 6.0-dev effective-5.10 (LLVM 04733f8187b260e, Swift 2744413e69179d8)'). Please select a toolchain which matches the SDK.
import SwiftUI
       ^
<unknown>:0: error: unexpected error produced: fatal error encountered while in -verify mode
/Applications/Xcode_16.0.0_16a5202i.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/System/Library/Frameworks/SwiftUICore.framework/Modules/SwiftUICore.swiftmodule/arm64e-apple-macos.swiftinterface:11018:41: error: diagnostic produced elsewhere: 'AnyShapeStyleBox' is only available in macOS 12.0 or newer; clients of 'SwiftUICore' may have a lower deployment target
package class AnyColorBox : SwiftUICore.AnyShapeStyleBox, @unchecked Swift.Sendable {
                                        ^
<unknown>:0: error: diagnostic produced elsewhere: fatal error encountered while in -verify mode
```

The same `swiftc` invocation works if I replace the just built `swift-frontend` path by `xcrun swift-frontend`.

### Reproduction

1. Install Xcode 16 beta 3.
2. `xcode-select -s` to Xcode 16 beta 3.
3. Use the `build-script` in Environment (or any build script)
4. Run the test suite and observe which tests fail.

### Expected behavior

The `.swiftinterface` files distributed with Xcode 16 SDKs should be valid for the open source `swiftc`. No error should be reported.

### Environment

Using `main` at 2744413e69179d8ee754184159acb62f09334ae9 from Tuesday July 9th, but I have been seeing this since early June and the first Xcode 16 beta and also in the `release/6.0` branch.

Currently using Xcode 16.0.0 beta 3 (16a5202i), but I have seen the problem with Beta 1 and Beta 2 as well.

My build script invocation is the following:

```
utils/build-script --build-subdir=my_macos --assertions --swift-enable-ast-verifier=0 --compiler-vendor=apple --sccache --release-debuginfo --swift-disable-dead-stripping --swift-enable-experimental-string-processing --test --validation-test --skip-test-swift --skip-test-libicu --skip-test-lldb --skip-test-foundation --skip-test-libdispatch --skip-test-xctest --skip-test-llbuild --skip-test-swiftpm --skip-test-cmark --skip-build-benchmarks --skip-build-tvos --skip-build-watchos --ios --maccatalyst
```

### Additional information

The following patch for the `swiftinterface` files seems to work around the problem:

```
@@ -5025,15 +5025,15 @@
 @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
 @_originallyDefinedIn(module: "SwiftUI", iOS 18.0)
 @_originallyDefinedIn(module: "SwiftUI", macOS 15.0)
 @_originallyDefinedIn(module: "SwiftUI", tvOS 18.0)
 @_originallyDefinedIn(module: "SwiftUI", watchOS 11.0)
 extension SwiftUICore.AnyShapeStyle.Storage : @unchecked Swift.Sendable {
 }
-@_hasMissingDesignatedInitializers @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+@_hasMissingDesignatedInitializers @available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
 @_originallyDefinedIn(module: "SwiftUI", iOS 18.0)
 @_originallyDefinedIn(module: "SwiftUI", macOS 15.0)
 @_originallyDefinedIn(module: "SwiftUI", tvOS 18.0)
 @_originallyDefinedIn(module: "SwiftUI", watchOS 11.0)
 @usableFromInline
 package class AnyShapeStyleBox {
   @objc @usableFromInline
```

`AnyShapeStyleBox` needs an `@available` compatible with the one in `AnyColorBox` (a subtype).

--- a/lib/Sema/TypeCheckAccess.cpp
+++ b/lib/Sema/TypeCheckAccess.cpp
@@ -2254,7 +2254,8 @@ public:
     // declaration slipped in when the compiler wasn't able to diagnose it and
     // can't be changed.
     if (nominal->getName().is("AnyColorBox") &&
-        nominal->getModuleContext()->getName().is("SwiftUI"))
+        (nominal->getModuleContext()->getName().is("SwiftUI") ||
+         nominal->getModuleContext()->getName().is("SwiftUICore")))
       flags |= DeclAvailabilityFlag::
           AllowPotentiallyUnavailableAtOrBelowDeploymentTarget;
 

